### PR TITLE
Backend implementation for installation assistent

### PR DESCRIPTION
**Pull request by mlang97**
### Features
* Collect data (e.g. information about installer, configuration about the edge)
* Create setup protocol (PDF) from collected data
* Register customer and installer via login page 
grafik
grafik

### CODE REVIEW

File: io.openems.backend.metadata.odoo/src/io/openems/backend/metadata/odoo/Field.java [hunk 1]
```diff
@@ -43,7 +44,7 @@ public static String getSqlQueryFields(Field[] fields) {
 		OPENEMS_SUM_STATE("openems_sum_state_level", true), //
 		OPENEMS_IS_CONNECTED("openems_is_connected", false);
 
-		public static final String ODOO_MODEL = "edge.device";
+		public static final String ODOO_MODEL = "fems.device";
```
Comment (sfeilmeier) [inline io.openems.backend.metadata.odoo/src/io/openems/backend/metadata/odoo/Field.java:47]: Please revert to edge.device

File: io.openems.backend.metadata.odoo/src/io/openems/backend/metadata/odoo/Field.java [hunk 2]
```diff
@@ -93,7 +95,7 @@ public boolean isQuery() {
 		LAST_ACKNOWLEDGE("last_acknowledge", false), //
 		ACKNOWLEDGE_DAYS("acknowledge_days", false);
 
-		public static final String ODOO_MODEL = "edge.device_status";
+		public static final String ODOO_MODEL = "fems.device_status";
```
Comment (sfeilmeier) [inline io.openems.backend.metadata.odoo/src/io/openems/backend/metadata/odoo/Field.java:98]: revert

File: io.openems.backend.metadata.odoo/src/io/openems/backend/metadata/odoo/Field.java [hunk 3]
```diff
@@ -139,7 +142,7 @@ public boolean isQuery() {
 		TEASER("teaser", false), //
 		DETAILS("details", false);
 
-		public static final String ODOO_MODEL = "edge.openemsconfigupdate";
+		public static final String ODOO_MODEL = "fems.openemsconfigupdate";
```
Comment (sfeilmeier) [inline io.openems.backend.metadata.odoo/src/io/openems/backend/metadata/odoo/Field.java:145]: revert

File: io.openems.backend.metadata.odoo/src/io/openems/backend/metadata/odoo/Field.java [hunk 4]
```diff
@@ -175,5 +178,400 @@ public int index() {
...
+
+	/**
+	 * The EdgeDeviceUserRole-Model.
+	 */
+	public enum EdgeDeviceUserRole implements Field {
+		DEVICE_ID("device_id", false), //
+		USER_ID("user_id", false), //
+		ROLE("role", false);
+
+		public static final String ODOO_MODEL = "fems.device_user_role";
```
Comment (sfeilmeier) [inline io.openems.backend.metadata.odoo/src/io/openems/backend/metadata/odoo/Field.java:192]: revert

File: io.openems.backend.metadata.odoo/src/io/openems/backend/metadata/odoo/Field.java [hunk 5]
```diff
@@ -175,5 +178,400 @@ public int index() {
...
+		}
+	}
+
+	public enum SetupProtocol implements Field {
+		CUSTOMER("customer_id", true), //
+		DIFFERENT_LOCATION("different_location_id", true), //
+		INSTALLER("installer_id", true), //
+		FEMS("fems_device_id", true);
+
+		public static final String ODOO_MODEL = "fems.setup_protocol";
```
Comment (sfeilmeier) [inline io.openems.backend.metadata.odoo/src/io/openems/backend/metadata/odoo/Field.java:397]: edge

File: io.openems.backend.metadata.odoo/src/io/openems/backend/metadata/odoo/Field.java [hunk 6]
```diff
@@ -175,5 +178,400 @@ public int index() {
...
+			return this.query;
+		}
+	}
+
+	public enum SetupProtocolProductionLot implements Field {
+		SETUP_PROTOCOL("setup_protocol_id", true), //
+		SEQUENCE("sequence", true), //
+		LOT("lot_id", true);
+
+		public static final String ODOO_MODEL = "fems.setup_protocol_production_lot";
```
Comment (sfeilmeier) [inline io.openems.backend.metadata.odoo/src/io/openems/backend/metadata/odoo/Field.java:444]: edge

File: io.openems.backend.metadata.odoo/src/io/openems/backend/metadata/odoo/Field.java [hunk 7]
```diff
@@ -175,5 +178,400 @@ public int index() {
...
+		public boolean isQuery() {
+			return this.query;
+		}
+	}
+
+	public enum SetupProtocolItem implements Field {
+		SETUP_PROTOCOL("setup_protocol_id", true), //
+		SEQUENCE("sequence", true);
+
+		public static final String ODOO_MODEL = "fems.setup_protocol_item";
```
Comment (sfeilmeier) [inline io.openems.backend.metadata.odoo/src/io/openems/backend/metadata/odoo/Field.java:490]: edge

File: io.openems.backend.metadata.odoo/src/io/openems/backend/metadata/odoo/odoo/OdooHandler.java [hunk 8]
```diff
@@ -68,6 +77,75 @@ public void addChatterMessage(MyEdge edge, String message) {
...
+					Field.EdgeDevice.NAME);
+
+			String name = (String) read.get(Field.EdgeDevice.NAME.id());
+			if (name == null) {
+				return Optional.empty();
+			}
+
+			return Optional.of(name);
+		} catch (OpenemsException e) {
+			this.parent.logInfo(this.log, "Unable to find Edge by setupPassowrd [" + setupPassword + "]");
```
Comment (sfeilmeier) [inline io.openems.backend.metadata.odoo/src/io/openems/backend/metadata/odoo/odoo/OdooHandler.java:105]: typo

File: ui/src/app/index/index.component.html [hunk 9]
```diff
@@ -58,10 +58,20 @@
...
                 <input type="submit" style="visibility: hidden; position: absolute; position: -999px" />
 
-                <ion-button type="submit" class="ion-float-right ion-margin-top ion-margin-bottom" size="medium">
-                  <small>LOGIN</small>
-                  <ion-icon slot="end" name="log-in-outline"></ion-icon>
-                </ion-button>
+                <ion-item lines="none" class="ion-margin-top">
+                  <table class=" full_width">
+                    <a type="button" class="ion-float-left" size="small"
+                      href="https://portal.fenecon.de/web/reset_password?" style="text-decoration: none;">
```
Comment (sfeilmeier) [inline ui/src/app/index/index.component.html:64]: to be removed

File: ui/src/app/registration/modal/modal.component.html [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+                <ion-item>
+                    <ion-label style="white-space: pre-wrap;" [innerHTML]="'Register.form.acceptAgb' | translate">
+                    </ion-label>
+                    <ion-checkbox slot="start" formControlName="acceptAgb"></ion-checkbox>
+                </ion-item>
+                <ion-item>
+                    <ion-label style="white-space: pre-wrap;"
+                        [innerHTML]="'Register.form.subscribeNewsletter' | translate">
+                    </ion-label>
+                    <ion-checkbox slot="start" formControlName="subscribeNewsletter"></ion-checkbox>
```
Comment (sfeilmeier) [inline ui/src/app/registration/modal/modal.component.html:183]: FENECON specific; should be removed

File: ui/src/app/shared/translate/de.ts [hunk 11]
```diff
@@ -112,8 +112,45 @@ export const TRANSLATION = {
...
     },
+    Register: {
+        title: "Benutzer Account anlegen",
+        segment: {
+            user: "Benutzer",
+            installer: "Installateur"
+        },
+        form: {
+            user: {
+                acceptPrivacyPolicy: "Durch das Erstellen eines FENECON Accounts erkläre ich, die FENECON <a target=\"_blank\" href=\"https://fenecon.de/page/datenschutzerklaerung/\">Datenschutzerklärung</a> und die Nutzungsbedingungen gelesen zu haben und stimme ihnen zu.*",
```
Comment (sfeilmeier) [inline ui/src/app/shared/translate/de.ts:126]: remove

**Review by sfeilmeier — CHANGES_REQUESTED** [review]
Some changes are required to comply with OpenEMS (instead of FEMS)

**Review by mlang97 — COMMENTED** [review]
All specified FEMS features are removed

**Review by sfeilmeier — APPROVED** [review]

**Pull request merged**